app/src/org/odk/collect/android/utilities/GeoUtils.java

```diff
@@ -56,6 +59,23 @@ public static String locationToString(Location location) {
         return set;
     }
 
+    public static Set<String> evaluateProvidersWithPermissions(LocationManager manager, Context context) {
+        HashSet<String> set = new HashSet<>();
+
+        List<String> providers = manager.getProviders(true);
+        for (String provider : providers) {
+            if ((provider.equalsIgnoreCase(LocationManager.GPS_PROVIDER) && ContextCompat.checkSelfPermission(
+                            context, Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED)) {
+                set.add(LocationManager.GPS_PROVIDER);
+            } if ((provider.equalsIgnoreCase(LocationManager.NETWORK_PROVIDER) && ContextCompat.checkSelfPermission(
+                            context, Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED)) {
+                set.add(LocationManager.NETWORK_PROVIDER);
+            }
+        }
```

Comment:
Yeah, I agree that it's a bit confusing, but I'll leave it to be consistent with the GeoUtils.evaluateProviders, which was already there.